rllib/models/tf/tf_action_dist.py

```diff
@@ -141,6 +142,70 @@ def required_model_output_shape(action_space, model_config):
         return np.sum(action_space.nvec)
 
 
+class GumbelSoftmax(TFActionDistribution):
+    """GumbelSoftmax distr. (for differentiable sampling in discr. actions
+
+    The Gumbel Softmax distribution [1] (also known as the Concrete [2]
+    distribution) is a close cousin of the relaxed one-hot categorical
+    distribution, whose tfp implementation we will use here plus
+    adjusted `sample_...` and `log_prob` methods. See discussion at [0].
+
+    [0] https://stackoverflow.com/questions/56226133/
+    soft-actor-critic-with-discrete-action-space
+
+    [1] Categorical Reparametrization with Gumbel-Softmax (Jang et al, 2017):
+    https://arxiv.org/abs/1611.01144
+    [2] The Concrete Distribution: A Continuous Relaxation of Discrete Random
+    Variables (Maddison et al, 2017) https://arxiv.org/abs/1611.00712
+    """
+
+    @DeveloperAPI
+    def __init__(self, inputs, model=None, temperature=1.0):
+        """Initializes a GumbelSoftmax distribution.
+
+        Args:
+            temperature (float): Temperature parameter. For low temperatures,
+                the expected value approaches a categorical random variable.
+                For high temperatures, the expected value approaches a uniform
+                distribution.
+        """
+        # Clamp for stability reasons.
+        temperature = max(0.0001, temperature)
```

Comment:
This seems a bit clumsy way of stabilizing the temperature. Would be nice to at least warn the user that the temperature being set to lower than `1e-4` is clamped.

Reply:
+1

Reply:
I replaced these with an assertion >= 0.0. Feel less arbitrary this way.